lib/auth/recordingmetadata/recordingmetadatav1/recordingmetadata.go

```diff
@@ -0,0 +1,350 @@
+	apievents "github.com/gravitational/teleport/api/types/events"
+	"github.com/gravitational/teleport/lib/events"
+	"github.com/gravitational/teleport/lib/player"
+	"github.com/gravitational/teleport/lib/session"
+	"github.com/gravitational/teleport/lib/terminal"
+)
+
+type uploadHandler interface {
+	// UploadMetadata uploads session metadata and returns a URL with the uploaded
+	// file in case of success. Session metadata is a file with a [recordingmetadatav1.SessionRecordingMetadata]
+	// protobuf message containing info about the session (duration, events, etc), as well as
+	// multiple [recordingmetadatav1.SessionRecordingThumbnail] messages (thumbnails).
+	UploadMetadata(ctx context.Context, sessionID session.ID, reader io.Reader) (string, error)
+	// UploadThumbnail uploads a session thumbnail and returns a URL with uploaded
+	// file in case of success. A thumbnail is [recordingmetadatav1.SessionRecordingThumbnail]
+	// protobuf message which contains the thumbnail as an SVG, and some basic details about the
+	// state of the terminal at the time of the thumbnail capture (terminal size, cursor position).
+	UploadThumbnail(ctx context.Context, sessionID session.ID, reader io.Reader) (string, error)
+}
+
+// RecordingMetadataService processes session recordings to generate metadata and thumbnails.
+type RecordingMetadataService struct {
+	logger        *slog.Logger
+	streamer      player.Streamer
+	uploadHandler uploadHandler
+}
+
+// RecordingMetadataServiceConfig defines the configuration for the RecordingMetadataService.
+type RecordingMetadataServiceConfig struct {
+	// Streamer is used to stream session events.
+	Streamer player.Streamer
+	// UploadHandler is used to upload session metadata and thumbnails.
+	UploadHandler uploadHandler
+}
+
+const (
+	// inactivityThreshold is the duration after which an inactivity event is recorded.
+	inactivityThreshold = 10 * time.Second
+
+	// maxThumbnails is the maximum number of thumbnails to store in the session metadata.
+	maxThumbnails = 1000
+)
+
+// NewRecordingMetadataService creates a new instance of RecordingMetadataService with the provided configuration.
+func NewRecordingMetadataService(cfg RecordingMetadataServiceConfig) (*RecordingMetadataService, error) {
+	if cfg.Streamer == nil {
+		return nil, trace.BadParameter("streamer is required")
+	}
+	if cfg.UploadHandler == nil {
+		return nil, trace.BadParameter("uploadHandler is required")
+	}
+
+	return &RecordingMetadataService{
+		streamer:      cfg.Streamer,
+		uploadHandler: cfg.UploadHandler,
+		logger:        slog.With(teleport.ComponentKey, "recording_metadata"),
+	}, nil
+}
+
+// ProcessSessionRecording processes the session recording associated with the provided session ID.
+// It streams session events, generates metadata, and uploads thumbnails and metadata.
+func (s *RecordingMetadataService) ProcessSessionRecording(ctx context.Context, sessionID session.ID) error {
```

Comment:
Nit: it definitely feels like this function is bordering on being too large/complex. I wonder if the setup or event handling could be extracted to distinct functions.